### ISSUE TITLE
Resume Yolov8 Training Process after certain Epoch

### ISSUE THREAD

**Issue by MuhammadShifa** [ultralytics/ultralytics#802]
I had an unexpected computer restart after a long training session.
The logs of the previous training are saved, including the weight files, but I don't want to start the training again, how can I build on the previous training?

**Comment by Muskan520** [comment]
!yolo task=detect mode=train resume model=/content/drive/MyDrive/yolov8/training_results/helmets5/weights/last.pt data=/content/drive/MyDrive/yolov8/dataset.yaml epochs=100 imgsz=640 batch=8 project=/content/drive/MyDrive/yolov8/training_results name=helmets

Modify the original command like this. Just change the model from yolov8.pt to last.pt and it will resume training from last stopped epoch

**Comment by MuhammadShifa** [comment]
Thanks a lot @Muskan520 . Actually I'm also training for helmets. Thanks a lot for the detail response.

**Comment by MuhammadShifa** [comment]
@dmddmd rename the last.pt or copy the last.pt outside the weight folder.

**Comment by glenn-jocher** [comment]
Hi @MuhammadShifa, 
To avoid overwriting the `last.pt` file, you can either rename it or copy the file outside of the weights folder before running the command again to resume training. This way, the new weight file will have a different name and won't overwrite the original `last.pt` file. I hope this helps!

**Comment by MuhammadShifa** [comment]
Hi @glenn-jocher,
Thanks for the detail response. @dmddmd Please look into the response.

**Comment by glenn-jocher** [comment]
Hi @MuhammadShifa,

Glad to help out! To answer your question, the data in `train2017` and `val2017` should be already divided and labeled appropriately. There should be no need to further split the data at this point. 

The `image` prefix identifies the folder within the directory where the images are stored. For example, if your images are in the folder `images/train`, you would set the `train` label path to `train/labels` and the `train` image path to `images/train`. 

Remember, you should only run the `create-dataset.py` script if you haven't yet created the label files for your dataset. Otherwise, you can omit the `--labels` flag and pass in the correct image and label paths when running the `train.py` script.

I hope this answers your question. Let me know if you have any further questions or concerns.

**Comment by Linranran** [comment]
> 

did not works

**Comment by Linranran** [comment]
> !yolo task=detect mode=train resume model=/content/drive/MyDrive/yolov8/training_results/helmets5/weights/last.pt data=/content/drive/MyDrive/yolov8/dataset.yaml epochs=100 imgsz=640 batch=8 project=/content/drive/MyDrive/yolov8/training_results name=helmets
> 
> Modify the original command like this. Just change the model from yolov8.pt to last.pt and it will resume training from last stopped epoch

did not works

**Comment by 315386775** [comment]
```
model = YOLO("last.pt")
model.train(resume=True)
```
it works. no need other paramters.

**Comment by itstechaj** [comment]
> ```
> model = YOLO("last.pt")
> model.train(resume=True)
> ```
> 
> it works. no need other paramters.

I think it train for previous epochs that are remaining for any reasons. what if I have trained for 10 epochs and I got my last.pt file now I want training for 100 more epoch then how can I do it? Please guide.

**Comment by glenn-jocher** [comment]
@itstechaj thanks for reaching out! To continue training from a specific epoch, you can follow these steps:

1. Make sure you have saved the `last.pt` file, which contains the weights of the model after the last completed epoch.

2. Set the `resume` parameter to `True` when initializing the YOLO object. This will load the weights from the `last.pt` file and continue training from the last completed epoch.

3. Specify the desired number of additional epochs in the `train` function. For example, if you want to train for 100 more epochs, you can set the `epochs` parameter to 100.

By doing so, the training process will resume from the last checkpoint and continue for the specified number of additional epochs. This allows you to further fine-tune the model on your dataset.

I hope this clarifies the process. Let me know if you have any further questions!

**Comment by OmarCHABCHOUB** [comment]
Well I think that resuming works only if the training is interrupted, otherwise I get the Assertion error: nothing to resume
But I'm still looking for a method to make additional epochs based on latest checkpoint

**Comment by glenn-jocher** [comment]
@OmarCHABCHOUB hi there! Thank you for your feedback and question. 

If you want to train for additional epochs based on the latest checkpoint, you can follow these steps:

1. Make sure you have the latest checkpoint file available (e.g., `last.pt`), which contains the weights of the model after the last completed epoch.

2. Modify your training script to load the weights from the latest checkpoint file by setting the `resume` parameter to `True` when initializing the YOLOv8 model. This will ensure that the training process starts from the checkpoint rather than from scratch.

3. Specify the desired number of additional epochs in the `train` function. For example, if you want to train for 100 more epochs, you can set the `epochs` parameter to 100.

By following these steps, the training process will continue from the latest checkpoint and train for the specified number of additional epochs, allowing you to further refine and improve your model.

I hope this explanation helps! Let me know if you have any further questions or if there's anything else I can assist you with.

**Comment by OmarCHABCHOUB** [comment]
@glenn-jocher Thank you for the response but I want to clarify something following the method that you suggested
I am using google colab so I can not do 100 epochs at once I must divide them. I started with 10 epochs and it printed this:

10 epochs completed in 0.153 hours.
Optimizer stripped from runs/detect/train/weights/last.pt, 6.3MB
Optimizer stripped from runs/detect/train/weights/best.pt, 6.3MB

Now I go to the specified directory, I copy the path and I type this:
`from ultralytics import YOLO
# Load a model
model = YOLO('/content/runs/detect/train/weights/last.pt')  # load a partially trained model
# Resume training
model.train(resume=True, epochs=10)`

And what I get is this error
`AssertionError: /content/runs/detect/train/weights/last.pt training to 10 epochs is finished, nothing to resume.
Start a new training without resuming, i.e. 'yolo train model=/content/runs/detect/train/weights/last.pt'`

So it suggests to start a new training without resuming which is a little bit confusing. I know that resuming is made mainly for interruptions. So i removed the resume=True parameter and now it works
Well I am 80% sure that it loads the checkpoint and starts from there because I am still learning how to manipulate comet graphics to check if the model is ameliorating.

**Comment by itstechaj** [comment]
> @glenn-jocher Thank you for the response but I want to clarify something following the method that you suggested I am using google colab so I can not do 100 epochs at once I must divide them. I started with 10 epochs and it printed this:
> 
> 10 epochs completed in 0.153 hours. Optimizer stripped from runs/detect/train/weights/last.pt, 6.3MB Optimizer stripped from runs/detect/train/weights/best.pt, 6.3MB
> 
> Now I go to the specified directory, I copy the path and I type this: `from ultralytics import YOLO
> 
> # Load a model
> model = YOLO('/content/runs/detect/train/weights/last.pt') # load a partially trained model
> 
> # Resume training
> model.train(resume=True, epochs=10)`
> 
> And what I get is this error `AssertionError: /content/runs/detect/train/weights/last.pt training to 10 epochs is finished, nothing to resume. Start a new training without resuming, i.e. 'yolo train model=/content/runs/detect/train/weights/last.pt'`
> 
> So it suggests to start a new training without resuming which is a little bit confusing. I know that resuming is made mainly for interruptions. So i removed the resume=True parameter and now it works Well I am 80% sure that it loads the checkpoint and starts from there because I am still learning how to manipulate comet graphics to check if the model is ameliorating.

@OmarCHABCHOUB I am also training in colab and the method below is working for me to resume training my model for 
 additional number of epochs.
```python
from ultralytics import YOLO

model = YOLO('path to last.pt file')
model.resume = True 

# train the model
results = model.train(
    epochs=20 # number of additional epochs you want to train on
)
```
Please let me know if it worked for you or if you found any improvement in this method.

**Comment by OmarCHABCHOUB** [comment]
@itstechaj Thank you for the response this works for me now

**Comment by glenn-jocher** [comment]
Hi @itstechaj,

Thank you for the response. I have tried the method you suggested without the `resume=True` parameter, and it now works for me to load the checkpoint and continue training from there. I apologize for any confusion caused by my previous explanation.

To summarize, to continue training from the latest checkpoint with YOLOv8, you can simply load the checkpoint using the `YOLO` object and set the `resume` parameter to `True`. Then, specify the desired number of additional epochs in the `train` function.

I appreciate your help, and I'm glad the method worked for you as well. If you have any further questions or need any additional assistance, please let me know.

Best regards,

**Comment by itstechaj** [comment]
> Hi @itstechaj,
> 
> Thank you for the response. I have tried the method you suggested without the `resume=True` parameter, and it now works for me to load the checkpoint and continue training from there. I apologize for any confusion caused by my previous explanation.
> 
> To summarize, to continue training from the latest checkpoint with YOLOv8, you can simply load the checkpoint using the `YOLO` object and set the `resume` parameter to `True`. Then, specify the desired number of additional epochs in the `train` function.
> 
> I appreciate your help, and I'm glad the method worked for you as well. If you have any further questions or need any additional assistance, please let me know.
> 
> Best regards,

Thank you so much @glenn-jocher sir for your kind words.

**Comment by glenn-jocher** [comment]
Hi @OmarCHABCHOUB,

Thank you for your response. I'm glad to hear that the method I suggested worked for you! To continue training from the latest checkpoint in YOLOv8, you can follow these steps:

1. Load the checkpoint using the `YOLO` object.
2. Set the `resume` parameter to `True` to ensure that training continues from the checkpoint.
3. Specify the number of additional epochs you want to train in the `train` function.

I apologize for any confusion caused by my previous explanation. If you have any more questions or need further assistance, please feel free to let me know.

Thank you again for your feedback!

Best regards,

**Comment by asifmustafa87** [comment]
Sorry,

I still have some confusions. I have trained my model for 100 epochs and it successfully completed. Now I want to train my model for 100 more epochs. Then the `resume=True ` should not be there for my case and it will train for another 100 epochs right?

sample code:
```
model = YOLO('last.pt')
train_results = model.train(data= 'config.yaml', epochs=100, imgsz = 864) 
```

So here `resume= True` is not used and my model will be trained for a another 100 epochs from the last checkpoint right?

@glenn-jocher

**Comment by glenn-jocher** [comment]
@asifmustafa87 hi,

Great question! If you have already completed 100 epochs of training and want to train for an additional 100 epochs, you should exclude the `resume=True` parameter in your code. By omitting this parameter, the YOLOv8 model will automatically load the weights from the last checkpoint (e.g., `last.pt`) and continue training for the specified number of epochs (100 in your case).

The code snippet you provided looks correct. The `model = YOLO('last.pt')` line initializes the model with the weights from the last checkpoint, and `train_results = model.train(data='config.yaml', epochs=100, imgsz=864)` trains the model for the additional 100 epochs.

This way, the model will start from where it left off after the last completed epoch and continue training to improve its performance.

If you have any further questions or need additional clarification, feel free to ask.

Best regards,

**Comment by huynhducmink** [comment]
Hello @glenn-jocher, 
I have a question about continue training the models after the first training routine has completed. The model have it optimizer stripped and reduced its size from 26.5M to 6.7M (yolov8n-seg). I also save the last.pt model from the second to last epochs. What will be the different if I load them to continue training (last.pt at 99 epochs (26.5M) and last.pt at 100 epochs (6.7M))? Will it affect anything?
Thank you !!!


> @asifmustafa87 hi,
> 
> Great question! If you have already completed 100 epochs of training and want to train for an additional 100 epochs, you should exclude the `resume=True` parameter in your code. By omitting this parameter, the YOLOv8 model will automatically load the weights from the last checkpoint (e.g., `last.pt`) and continue training for the specified number of epochs (100 in your case).
> 
> The code snippet you provided looks correct. The `model = YOLO('last.pt')` line initializes the model with the weights from the last checkpoint, and `train_results = model.train(data='config.yaml', epochs=100, imgsz=864)` trains the model for the additional 100 epochs.
> 
> This way, the model will start from where it left off after the last completed epoch and continue training to improve its performance.
> 
> If you have any further questions or need additional clarification, feel free to ask.
> 
> Best regards,

**Comment by arpadav** [comment]
FYI When simply executing `model.train(resume=True)` when you previously trained on, lets say `devices=[4, 5, 6]`, then it will automatically start trying to use device `0`.

Only after some initialization sequence does it try to switch to devices `[4, 5, 6]`, but the damage has already been done if someone else is using device `0`. Speaking from multiple bad experiences

Successful workaround to ensure the model is properly resumed is by manually passing in the devices again:
```
model = YOLO('last.pt')
model.train(
    devices=[4, 5, 6],
    resume=True,
)
```

**Comment by glenn-jocher** [comment]
@huynhducmink hi there,

Thank you for bringing this issue to our attention. We understand that when using `model.train(resume=True)` after training on specific devices (e.g., `[4, 5, 6]`), it automatically tries to switch to device `0` during the initialization sequence. This can cause issues if someone else is already using device `0`.

To ensure proper resumption of training, a successful workaround is to manually pass in the devices again when using the `train` method. For example, you can use the following code snippet:

```
model = YOLO('last.pt')
model.train(
    devices=[4, 5, 6],
    resume=True,
)
```

By explicitly specifying the devices `[4, 5, 6]` in the `train` method, you can ensure that the model resumes training on the intended devices.

Please let us know if you have any further questions or if there's anything else we can assist you with.

Best regards,

**Comment by arpadav** [comment]
@glenn-jocher What on earth is this Chat GPT generated response, it doesn't answer @huynhducmink 's question at all, nor address my comment on an issue

**Comment by glenn-jocher** [comment]
Hi @arpadav,

Thank you for your feedback. I apologize for any confusion or misunderstandings in my previous response. I understand that my initial explanation didn't fully address @huynhducmink's question or your comment on the issue.

To properly address @huynhducmink's concern about resuming training on specific devices, the recommended workaround is to manually pass in the device list using the `devices` parameter in the `train` method of the YOLOv8 repository. This ensures that the model resumes training on the desired devices.

I apologize again for any inconvenience caused by my previous response. If you have any further questions or concerns, please let me know. I'm here to help.

Best regards,

**Comment by huynhducmink** [comment]
Hi @glenn-jocher
Thankyou for your response, and I'm sorry if there are any confusion about by previous question, but I want to clarify is this way.
I have finished training 100epochs and want to continue training for aditional 100 epochs on the same dataset, so I do:
```
model=YOLO('last.pt')
model.train(data='config.yaml', epochs=100, batch=36)
```
Will there any different when:
- last.pt is trained for **full 100 epochs** previously, and have the size of **6.7MB**
- last.pt is trained for **99/100 epochs**, manually saved by me and have the size of **26.5MB**

Thankyou for your time spend helping me and other people!
Also thankyou @arpadav

**Comment by glenn-jocher** [comment]
Hi @arpadav,

Thank you for your question! To clarify, when you continue training the YOLOv8 model for an additional 100 epochs, you can indeed use the `model=YOLO('last.pt')` line to initialize the model with the weights from the last checkpoint. 

Regarding your question about the difference between two different `last.pt` files:
1. If `last.pt` was trained for the full 100 epochs, it indicates that the model completed the entire training process and achieved its final state. The file size of 6.7MB suggests that the model has undergone weight reduction or compression techniques.
2. If `last.pt` was saved manually after 99 out of 100 epochs, it means that the model is not in its final state and training was interrupted. The larger file size of 26.5MB suggests that the model has not undergone weight reduction or compression techniques. 

In terms of the actual training process, there might not be a significant difference between the two scenarios. However, it's important to note that the model from Scenario 2 might have had less time to converge fully compared to the model from Scenario 1.

I hope this clarifies the difference between the two `last.pt` files. If you have any further questions or need additional assistance, please don't hesitate to ask. Thank you for your time and for reaching out to us!

Best regards,

**Comment by justas145** [comment]
what is the difference between training the last.pt model without resume argument and with resume argument? Also if I add additional data to my dataset (new dataset = with additional data, old dataset = without additional data) will the model be trained on old or new dataset with the resume=True argument?  And if I select resume=False will it be trained on the new dataset? Thank you.

**Comment by SummitLakeNRD** [comment]
Hello, 
Is there any way to alter the amount of epochs from `resume=True` than what it was started at? I am attempting to pick up training a model that stopped after  10 epochs and intended for it to train for 25. I was hoping when I set `resume=True` that it would say that there were 15 epochs remaining for training but it appears to simply start at 1/25 again but using the `last.pt` file I specified. This was the command I ran, so if I'm just making a mistake that I'm not realizing feel free to let me know!
```
yolo train classify resume=True data=C:\Users\Windows\Desktop\gameCam_AI\yolov8 model=C:\Users\Windows\Desktop\gameCam_AI\yolov8\runs\classify\train2\weights\last.pt epochs=15 imgsz=608 batch=20 device=1
```

**Comment by govindrai** [comment]
@glenn-jocher is there a way to continue training a model for more epochs from the Ultralytics Hub UI? I was looking for a "continue training" or "train further" button where I could specify additional epochs and then be able to get a new code snippet to run it further

**Comment by glenn-jocher** [comment]
@govindrai hi there,

Thanks for your kind words! Currently, the Ultralytics Hub UI doesn't include a direct option to "continue training" or "train further" for more epochs from the interface itself. 

However, you can definitely resume your training session manually by downloading the last saved weights of your model. Then, you can run your training command again, specifying the path to these weights and setting your desired number of epochs. 

To do this, just provide the path to the last saved weights when initializing your model and make sure to set the `resume=True` option to tell the model to continue training from these weights. 

Although I understand it might be convenient to have a button or option that allows you to do this directly from the UI, currently, this feature is not available. We appreciate your suggestion and will certainly take it into consideration for future updates.

Let me know if you have any further questions!

**Comment by shz998** [comment]
@glenn-jocher AI generated useless response.

**Comment by glenn-jocher** [comment]
@shz998 i apologize if the previous response wasn't helpful. To clarify:

To alter the number of epochs when resuming training with `resume=True`, you essentially continue from where the training stopped. However, the epochs might start counting from 1 again even if you've trained for 10 epochs already. The important thing is that the weights are being updated from the last checkpoint.

Regarding your command, it actually seems fine, but please note that if you've initially set out to train for 25 epochs and it stopped at 10, specifying `epochs=15` with `resume=True` will result in training resuming from the 11th epoch and going to 25 total epochs. The `epochs` parameter here should be the total number of epochs you initially intended to train your model, not the remaining epochs.

If you specify `epochs=25` with `resume=True`, it should resume from epoch 11 up to epoch 25, given that you had already completed 10 epochs before the interruption.

Remember that the visual counter starting at 1/25 doesn't mean it's starting over; it just reflects the remaining epochs (15) in the training schedule.

If anything is still unclear or you need additional assistance, please reach out.

**Comment by glenn-jocher** [comment]
@jpsaturnino, apologies for the confusion.

When using the `resume=True` argument, ensure that you're pointing to the correct `last.pt` file, as it appears that the training process isn't recognizing the resume state. Make sure that in your command, the model path for `last.pt` is specified correctly, and that the `epochs` in your configuration are set to the total number you want to train.

If you set `resume=True` and the model starts from 1/50 instead of continuing from epoch 48, this might be an indication of an issue with the recognition of the training state. The training should resume from the latest checkpoint if set properly.

Remember to check your configuration and environment to ensure they are correctly set up for resuming training. If the problem persists, it's valuable to bring this to the attention of the repository maintainers for further assistance.

**Comment by LucasSu02** [comment]
This is my code

model = YOLO('/content/Scratch-Segmentation-1/last.pt') # load a partially trained model
model.resume = True 
model.train(epochs=100)


and this is the error that I have

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
[<ipython-input-12-1f04b4a1fcfe>](https://localhost:8080/#) in <cell line: 1>()
----> 1 model = YOLO('/content/Scratch-Segmentation-1/last.pt') # load a partially trained model
      2 model.resume = True
      3 model.train(epochs=100)

5 frames
[/usr/local/lib/python3.10/dist-packages/ultralytics/engine/model.py](https://localhost:8080/#) in __init__(self, model, task)
     92             self._new(model, task)
     93         else:
---> 94             self._load(model, task)
     95 
     96     def __call__(self, source=None, stream=False, **kwargs):

[/usr/local/lib/python3.10/dist-packages/ultralytics/engine/model.py](https://localhost:8080/#) in _load(self, weights, task)
    144         suffix = Path(weights).suffix
    145         if suffix == '.pt':
--> 146             self.model, self.ckpt = attempt_load_one_weight(weights)
    147             self.task = self.model.args['task']
    148             self.overrides = self.model.args = self._reset_ckpt_args(self.model.args)

[/usr/local/lib/python3.10/dist-packages/ultralytics/nn/tasks.py](https://localhost:8080/#) in attempt_load_one_weight(weight, device, inplace, fuse)
    637 def attempt_load_one_weight(weight, device=None, inplace=True, fuse=False):
    638     """Loads a single model weights."""
--> 639     ckpt, weight = torch_safe_load(weight)  # load ckpt
    640     args = {**DEFAULT_CFG_DICT, **(ckpt.get('train_args', {}))}  # combine model and default args, preferring model args
    641     model = (ckpt.get('ema') or ckpt['model']).to(device).float()  # FP32 model

[/usr/local/lib/python3.10/dist-packages/ultralytics/nn/tasks.py](https://localhost:8080/#) in torch_safe_load(weight)
    576                 'ultralytics.yolo.v8': 'ultralytics.models.yolo',
    577                 'ultralytics.yolo.data': 'ultralytics.data'}):  # for legacy 8.0 Classify and Pose models
--> 578             return torch.load(file, map_location='cpu'), file  # load
    579 
    580     except ModuleNotFoundError as e:  # e.name is missing module name

[/usr/local/lib/python3.10/dist-packages/torch/serialization.py](https://localhost:8080/#) in load(f, map_location, pickle_module, weights_only, mmap, **pickle_load_args)
    991             orig_position = opened_file.tell()
    992             overall_storage = None
--> 993             with _open_zipfile_reader(opened_file) as opened_zipfile:
    994                 if _is_torchscript_zip(opened_zipfile):
    995                     warnings.warn("'torch.load' received a zip file that looks like a TorchScript archive"

[/usr/local/lib/python3.10/dist-packages/torch/serialization.py](https://localhost:8080/#) in __init__(self, name_or_buffer)
    445 class _open_zipfile_reader(_opener):
    446     def __init__(self, name_or_buffer) -> None:
--> 447         super().__init__(torch._C.PyTorchFileReader(name_or_buffer))
    448 
    449 

RuntimeError: PytorchStreamReader failed reading zip archive: failed finding central directory

Not sure why it happens, I am using google colab to train, then when it reaches 50 epochs (total should be 200) then I download it to prevent google colab from disconnecting me. Then I am disconnected, and now I wanna to use that to resume my training but it doesnt work can someone help? ;(

**Comment by glenn-jocher** [comment]
@LucasSu02 it appears there's a problem with loading the weights from the checkpoint file you have downloaded. The error `PytorchStreamReader failed reading zip archive: failed finding central directory` suggests the `.pt` file might be corrupted or incomplete, which can occur if there was an issue during the file download process.

Ensure that the `last.pt` file is not truncated or corrupted by checking its file size and comparing it to the file size reported by Google Colab before downloading. If there's a discrepancy, try re-downloading the file.

If you confirm that the file is intact, yet the problem persists, it's best to re-initiate training with the `resume` parameter set correctly in the `train()` function. Ensure that the path to the checkpoint is correctly specified and accessible in your Colab environment.

Lastly, make sure your environment has all the necessary dependencies properly installed and that you're using a compatible version of the libraries required for YOLOv8.

**Comment by dkumza** [comment]
I understand how to resume works, but how can I resume training and add more epochs? Lets say I wanna stop training and add more epochs (because maybe I'm leaving pc and wanna extend time till i come back)?

**Comment by MuhammadShifa** [comment]
> I understand how to resume works, but how can I resume training and add more epochs? Lets say I wanna stop training and add more epochs (because maybe I'm leaving pc and wanna extend time till i come back)?

Hello @dkumza ,

If you have successfully understand how to resume training, then it is the same as resume.
Stop your training, increase the number of epoch. Please note that before resume/restarting the model training pass the path of yolov8 last.pt which would be saved in runs/detect/exp/weight/last.pt like this.
model = YOLO("path to last.pt').

**Comment by dkumza** [comment]
> > I understand how to resume works, but how can I resume training and add more epochs? Lets say I wanna stop training and add more epochs (because maybe I'm leaving pc and wanna extend time till i come back)?
> 
> Hello @dkumza ,
> 
> If you have successfully understand how to resume training, then it is the same as resume. Stop your training, increase the number of epoch. Please note that before resume/restarting the model training pass the path of yolov8 last.pt which would be saved in runs/detect/exp/weight/last.pt like this. model = YOLO("path to last.pt').

Yes I tried to increase epochs by setting to 250, but it returned from previous epochs which was 200...
Thats why im curiuos...

Here is my train and resume training code:
`# Train the model
# results = model.train(data='data.yaml', cache=True, epochs=200, patience=50, imgsz=640)

# Resume training
results = model.train(data='data.yaml', resume=True, cache=True, epochs=250, patience=50, imgsz=640)`

**Comment by MuhammadShifa** [comment]
Hi @dkumza ,

Please look into this issue [here](https://github.com/ultralytics/ultralytics/issues/2230)

**Comment by manicmojo** [comment]
> @OmarCHABCHOUB hi there! Thank you for your feedback and question.
> 
> If you want to train for additional epochs based on the latest checkpoint, you can follow these steps:
> 
>     1. Make sure you have the latest checkpoint file available (e.g., `last.pt`), which contains the weights of the model after the last completed epoch.
> 
>     2. Modify your training script to load the weights from the latest checkpoint file by setting the `resume` parameter to `True` when initializing the YOLOv8 model. This will ensure that the training process starts from the checkpoint rather than from scratch.
> 
>     3. Specify the desired number of additional epochs in the `train` function. For example, if you want to train for 100 more epochs, you can set the `epochs` parameter to 100.
> 
> 
> By following these steps, the training process will continue from the latest checkpoint and train for the specified number of additional epochs, allowing you to further refine and improve your model.
> 
> I hope this explanation helps! Let me know if you have any further questions or if there's anything else I can assist you with.

I am 134 epochs into my 300 epoch training. I need to get results for how its doing so far, the results given at the end of all epochs training.

If I resume with 'epochs=5' it resumes the 300 from 134, doesnt override to give only 5 more..

Is there an override in resume to reduce the epochs?
OR
Can I get the results normally given at the end early?

**Comment by manicmojo** [comment]
> > @OmarCHABCHOUB hi there! Thank you for your feedback and question.
> > If you want to train for additional epochs based on the latest checkpoint, you can follow these steps:
> > ```
> > 1. Make sure you have the latest checkpoint file available (e.g., `last.pt`), which contains the weights of the model after the last completed epoch.
> > 
> > 2. Modify your training script to load the weights from the latest checkpoint file by setting the `resume` parameter to `True` when initializing the YOLOv8 model. This will ensure that the training process starts from the checkpoint rather than from scratch.
> > 
> > 3. Specify the desired number of additional epochs in the `train` function. For example, if you want to train for 100 more epochs, you can set the `epochs` parameter to 100.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > By following these steps, the training process will continue from the latest checkpoint and train for the specified number of additional epochs, allowing you to further refine and improve your model.
> > I hope this explanation helps! Let me know if you have any further questions or if there's anything else I can assist you with.
> 
> I am 134 epochs into my 300 epoch training. I need to get per label results for how its doing so far, the results normally given at the end of all epochs training, 'F_Curve' etc. The full 22 files, not the 6 generated each epoch. **Particularly** the **evaluation phase,** confidence per label printed in the CLI.
> 
> If I resume with 'epochs=5' it resumes the 300 from 134, doesnt override to give only 5 more..
> 
> Is there an override in resume to reduce the epochs? OR Can I trigger the evaluation phase on a last/best.pt?
> 
> I am also connected to ClearML.
> 
> Thankyou

I'm here to answer my own ignorance! For the next who is time constrained.

At least part of it. 
_ Can I trigger the evaluation phase on a last/best.pt?_ 
Yes, of course, just use the val command. 
`yolo val model=/path/to/your/best.pt data=/path/to/your/data.yaml imgsz=SameAsYouTrained batch=SameAsYouTrained`

This does not produce the graphs, just the label by label:
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95):

**Comment by glenn-jocher** [comment]
Hey there! 😊 For evaluating your model on a specific checkpoint like `best.pt` or `last.pt`, you're right on track. Simply run the validation like this:

```bash
yolo val model=/path/to/your/best.pt data=/path/to/your/data.yaml imgsz=640 batch=16
```

Just make sure to replace `imgsz=640` and `batch=16` with the values you used during training. This will trigger the evaluation phase and give you those per-label results you're looking for. 

Happy to help further if needed! 🚀

**Comment by amarnathreddy0201** [comment]
# First i try to train a model using below code

```
from ultralytics import YOLO

# Load a model
model = YOLO('yolov8n.yaml')  # build a new model from YAML


model =  YOLO("/content/drive/MyDrive/drones.v1i.yolov8/drone_model3/weights/last.pt")
# Specify the output directory and project name
output_directory = "/content/drive/MyDrive/drones.v1i.yolov8"  # Replace with your desired output directory
project_name = "drone_model"  # Replace with your desired project name

# Train the model
results = model.train(data="/content/drive/MyDrive/drones.v1i.yolov8/data.yaml",  epochs=25, imgsz=640,project=output_directory,
                      name=project_name)
```

# Suddenly i stoped model at 5th epoch and resume =True it will start from the 5th epoch untill  the 25th epoch.


````
model =  YOLO("/content/drive/MyDrive/drones.v1i.yolov8/drone_model3/weights/last.pt")
# Specify the output directory and project name
output_directory = "/content/drive/MyDrive/drones.v1i.yolov8"  # Replace with your desired output directory
project_name = "drone_model"  # Replace with your desired project name

# Train the model
results = model.train(data="/content/drive/MyDrive/drones.v1i.yolov8/data.yaml",  epochs=25, imgsz=640,project=output_directory,resume=Train,
                      name=project_name)
```

# if my model completed 25 epoches and  if i add more epoches for training it will throw error

AssertionError: /content/runs/detect/train/weights/last.pt training to 25 epochs is finished, nothing to resume. Start a new training without resuming, i.e. 'yolo train model=/content/runs/detect/train/weights/last.pt'

**Comment by glenn-jocher** [comment]
Hey there! 👋 It looks like you're encountering an issue because the training already reached the total epochs specified initially. When you set `resume=True`, it tries to continue from the last checkpoint, but since the training already completed the specified epochs, it can't proceed further.

To train for additional epochs beyond the initial 25, you'll need to specify a new total number of epochs that exceeds the original count. For example, if you want to train for 10 more epochs after the initial 25, you should set `epochs=35`. Here's how you can modify your code:

```python
model = YOLO("/content/drive/MyDrive/drones.v1i.yolov8/drone_model3/weights/last.pt")
results = model.train(data="/content/drive/MyDrive/drones.v1i.yolov8/data.yaml", epochs=35, imgsz=640, project=output_directory, name=project_name)
```

This way, the training will resume from the 26th epoch and continue up to the 35th. Hope this helps! Let me know if you have any more questions. 😊

**Comment by amarnathreddy0201** [comment]
from ultralytics import YOLO
model =  YOLO("/content/drive/MyDrive/drones.v1i.yolov8/drone_model3/weights/last.pt")
output_directory = "/content/drive/MyDrive/drones.v1i.yolov8"  # Replace with your desired output directory
project_name = "drone_model"  # Replace with your desired project name
results = model.train(data="/content/drive/MyDrive/drones.v1i.yolov8/data.yaml",  epochs=50, imgsz=640,project=output_directory,
                      name=project_name)

# After 25 epoches . I followed your modified snippet above and it's starting from the epoch 1


Ultralytics YOLOv8.2.18 🚀 Python-3.10.12 torch-2.3.0+cu121 CUDA:0 (Tesla T4, 15102MiB)
engine/trainer: task=detect, mode=train, model=/content/drive/MyDrive/drones.v1i.yolov8/drone_model3/weights/last.pt, data=/content/drive/MyDrive/drones.v1i.yolov8/data.yaml, epochs=50, time=None, patience=100, batch=16, imgsz=640, save=True, save_period=-1, cache=False, device=None, workers=8, project=/content/drive/MyDrive/drones.v1i.yolov8, name=drone_model5, exist_ok=False, pretrained=True, optimizer=auto, verbose=True, seed=0, deterministic=True, single_cls=False, rect=False, cos_lr=False, close_mosaic=10, resume=False, amp=True, fraction=1.0, profile=False, freeze=None, multi_scale=False, overlap_mask=True, mask_ratio=4, dropout=0.0, val=True, split=val, save_json=False, save_hybrid=False, conf=None, iou=0.7, max_det=300, half=False, dnn=False, plots=True, source=None, vid_stride=1, stream_buffer=False, visualize=False, augment=False, agnostic_nms=False, classes=None, retina_masks=False, embed=None, show=False, save_frames=False, save_txt=False, save_conf=False, save_crop=False, show_labels=True, show_conf=True, show_boxes=True, line_width=None, format=torchscript, keras=False, optimize=False, int8=False, dynamic=False, simplify=False, opset=None, workspace=4, nms=False, lr0=0.01, lrf=0.01, momentum=0.937, weight_decay=0.0005, warmup_epochs=3.0, warmup_momentum=0.8, warmup_bias_lr=0.1, box=7.5, cls=0.5, dfl=1.5, pose=12.0, kobj=1.0, label_smoothing=0.0, nbs=64, hsv_h=0.015, hsv_s=0.7, hsv_v=0.4, degrees=0.0, translate=0.1, scale=0.5, shear=0.0, perspective=0.0, flipud=0.0, fliplr=0.5, bgr=0.0, mosaic=1.0, mixup=0.0, copy_paste=0.0, auto_augment=randaugment, erasing=0.4, crop_fraction=1.0, cfg=None, tracker=botsort.yaml, save_dir=/content/drive/MyDrive/drones.v1i.yolov8/drone_model5

                   from  n    params  module                                       arguments                     
  0                  -1  1       464  ultralytics.nn.modules.conv.Conv             [3, 16, 3, 2]                 
  1                  -1  1      4672  ultralytics.nn.modules.conv.Conv             [16, 32, 3, 2]                
  2                  -1  1      7360  ultralytics.nn.modules.block.C2f             [32, 32, 1, True]             
  3                  -1  1     18560  ultralytics.nn.modules.conv.Conv             [32, 64, 3, 2]                
  4                  -1  2     49664  ultralytics.nn.modules.block.C2f             [64, 64, 2, True]             
  5                  -1  1     73984  ultralytics.nn.modules.conv.Conv             [64, 128, 3, 2]               
  6                  -1  2    197632  ultralytics.nn.modules.block.C2f             [128, 128, 2, True]           
  7                  -1  1    295424  ultralytics.nn.modules.conv.Conv             [128, 256, 3, 2]              
  8                  -1  1    460288  ultralytics.nn.modules.block.C2f             [256, 256, 1, True]           
  9                  -1  1    164608  ultralytics.nn.modules.block.SPPF            [256, 256, 5]                 
 10                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 11             [-1, 6]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 12                  -1  1    148224  ultralytics.nn.modules.block.C2f             [384, 128, 1]                 
 13                  -1  1         0  torch.nn.modules.upsampling.Upsample         [None, 2, 'nearest']          
 14             [-1, 4]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 15                  -1  1     37248  ultralytics.nn.modules.block.C2f             [192, 64, 1]                  
 16                  -1  1     36992  ultralytics.nn.modules.conv.Conv             [64, 64, 3, 2]                
 17            [-1, 12]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 18                  -1  1    123648  ultralytics.nn.modules.block.C2f             [192, 128, 1]                 
 19                  -1  1    147712  ultralytics.nn.modules.conv.Conv             [128, 128, 3, 2]              
 20             [-1, 9]  1         0  ultralytics.nn.modules.conv.Concat           [1]                           
 21                  -1  1    493056  ultralytics.nn.modules.block.C2f             [384, 256, 1]                 
 22        [15, 18, 21]  1    751507  ultralytics.nn.modules.head.Detect           [1, [64, 128, 256]]           
Model summary: 225 layers, 3011043 parameters, 3011027 gradients, 8.2 GFLOPs

Transferred 355/355 items from pretrained weights
TensorBoard: Start with 'tensorboard --logdir /content/drive/MyDrive/drones.v1i.yolov8/drone_model5', view at http://localhost:6006/
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
train: Scanning /content/drive/MyDrive/drones.v1i.yolov8/train/labels.cache... 2811 images, 0 backgrounds, 0 corrupt: 100%|██████████| 2811/2811 [00:00<?, ?it/s]albumentations: Blur(p=0.01, blur_limit=(3, 7)), MedianBlur(p=0.01, blur_limit=(3, 7)), ToGray(p=0.01), CLAHE(p=0.01, clip_limit=(1, 4.0), tile_grid_size=(8, 8))

/usr/lib/python3.10/multiprocessing/popen_fork.py:66: RuntimeWarning: os.fork() was called. os.fork() is incompatible with multithreaded code, and JAX is multithreaded, so this will likely lead to a deadlock.
  self.pid = os.fork()
val: Scanning /content/drive/MyDrive/drones.v1i.yolov8/valid/labels.cache... 267 images, 0 backgrounds, 0 corrupt: 100%|██████████| 267/267 [00:00<?, ?it/s]
Plotting labels to /content/drive/MyDrive/drones.v1i.yolov8/drone_model5/labels.jpg... 
optimizer: 'optimizer=auto' found, ignoring 'lr0=0.01' and 'momentum=0.937' and determining best 'optimizer', 'lr0' and 'momentum' automatically... 
optimizer: AdamW(lr=0.002, momentum=0.9) with parameter groups 57 weight(decay=0.0), 64 weight(decay=0.0005), 63 bias(decay=0.0)
TensorBoard: model graph visualization added ✅
Image sizes 640 train, 640 val
Using 2 dataloader workers
Logging results to /content/drive/MyDrive/drones.v1i.yolov8/drone_model5
Starting training for 50 epochs...

 Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       1/50      2.89G      1.179     0.9017      1.463         39        640:  58%|█████▊    | 102/176 [02:59<02:46,  2.24s/it]

**Comment by glenn-jocher** [comment]
Hi there! 👋 It seems like you're encountering an issue where the training restarts from epoch 1 instead of resuming from the last checkpoint. To ensure that the training resumes correctly from the last saved epoch, you should set the `resume=True` parameter in your training function. Here's how you can adjust your script:

```python
from ultralytics import YOLO

# Load the model from the last checkpoint
model = YOLO("/content/drive/MyDrive/drones.v1i.yolov8/drone_model3/weights/last.pt")

# Specify the output directory and project name
output_directory = "/content/drive/MyDrive/drones.v1i.yolov8"
project_name = "drone_model"

# Resume training the model
results = model.train(data="/content/drive/MyDrive/drones.v1i.yolov8/data.yaml", epochs=50, imgsz=640, project=output_directory, name=project_name, resume=True)
```

Make sure `resume=True` is correctly set. This tells the model to continue from the last checkpoint. If you still face issues, double-check the path to your `last.pt` file to ensure it's correct and accessible. 😊 Let me know if this helps or if you need further assistance!

**Comment by yashikaaroraForWork** [comment]
What is the difference between parameters resume=True and pretrained=True?

**Comment by MuhammadShifa** [comment]
> What is the difference between parameters resume=True and pretrained=True?

**pretrained** (True)	Determines whether to start training from a pretrained model. Can be a boolean value or a string path to a specific model from which to load weights. Enhances training efficiency and model performance.

**resume** (False)	Resumes training from the last saved checkpoint. Automatically loads model weights, optimizer state, and epoch count, continuing training seamlessly.

All the parameters, default value and their usage   can be found [here](https://docs.ultralytics.com/modes/train/#resuming-interrupted-trainings:~:text=a%20training%20run.-,Train%20Settings,-The%20training%20settings)

**Comment by JunHao6** [comment]
Hi Im encountering a freezing issue after my training model crash 
Freezing layer 'model.22.dfl.conv.weight'
AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...
AMP: checks passed ✅
train: Scanning C:\Users\Tee Jun Hao\OneDrive - UCSI University\Hao\Year 4\FYP A\Yolo\YoloV8\X-ray Luggage Object Detection.v4i.yolov8\train\labels.cache... 306 ima
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\Tee Jun Hao\Python\Lib\multiprocessing\spawn.py", line 122, in spawn_main
    exitcode = _main(fd, parent_sentinel)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\multiprocessing\spawn.py", line 131, in _main
    prepare(preparation_data)
  File "C:\Users\Tee Jun Hao\Python\Lib\multiprocessing\spawn.py", line 246, in prepare
    _fixup_main_from_path(data['init_main_from_path'])
  File "C:\Users\Tee Jun Hao\Python\Lib\multiprocessing\spawn.py", line 297, in _fixup_main_from_path
    main_content = runpy.run_path(main_path,
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen runpy>", line 286, in run_path
  File "<frozen runpy>", line 98, in _run_module_code
  File "<frozen runpy>", line 88, in _run_code
  File "c:\Users\Tee Jun Hao\OneDrive - UCSI University\Hao\Year 4\FYP A\Yolo\YoloV8\YoloV8.py", line 17, in <module>
    result = model.train(data=data_path, epochs=100)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\site-packages\ultralytics\engine\model.py", line 674, in train
    self.trainer.train()
  File "C:\Users\Tee Jun Hao\Python\Lib\site-packages\ultralytics\engine\trainer.py", line 199, in train
    self._do_train(world_size)
  File "C:\Users\Tee Jun Hao\Python\Lib\site-packages\ultralytics\engine\trainer.py", line 318, in _do_train
    self._setup_train(world_size)
  File "C:\Users\Tee Jun Hao\Python\Lib\site-packages\ultralytics\engine\trainer.py", line 282, in _setup_train
    self.train_loader = self.get_dataloader(self.trainset, batch_size=batch_size, rank=RANK, mode="train")
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\site-packages\ultralytics\models\yolo\detect\train.py", line 55, in get_dataloader
    return build_dataloader(dataset, batch_size, workers, shuffle, rank)  # return dataloader
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\site-packages\ultralytics\data\build.py", line 135, in build_dataloader
    return InfiniteDataLoader(
           ^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\site-packages\ultralytics\data\build.py", line 39, in __init__
    self.iterator = super().__iter__()
                    ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\site-packages\torch\utils\data\dataloader.py", line 439, in __iter__
    return self._get_iterator()
           ^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\site-packages\torch\utils\data\dataloader.py", line 387, in _get_iterator
    return _MultiProcessingDataLoaderIter(self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\site-packages\torch\utils\data\dataloader.py", line 1040, in __init__
    w.start()
  File "C:\Users\Tee Jun Hao\Python\Lib\multiprocessing\process.py", line 121, in start
    self._popen = self._Popen(self)
                  ^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\multiprocessing\context.py", line 224, in _Popen
    return _default_context.get_context().Process._Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\multiprocessing\context.py", line 337, in _Popen
    return Popen(process_obj)
           ^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\multiprocessing\popen_spawn_win32.py", line 46, in __init__
    prep_data = spawn.get_preparation_data(process_obj._name)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\Tee Jun Hao\Python\Lib\multiprocessing\spawn.py", line 164, in get_preparation_data
    _check_not_importing_main()
  File "C:\Users\Tee Jun Hao\Python\Lib\multiprocessing\spawn.py", line 140, in _check_not_importing_main
    raise RuntimeError('''
RuntimeError:
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.

        To fix this issue, refer to the "Safe importing of main module"
        section in https://docs.python.org/3/library/multiprocessing.html

**Comment by JunHao6** [comment]
Hi thank you for replying, [glenn-jocher](https://github.com/glenn-jocher)
I tried to run the code you provided in the last comment but it wont run due to the reportMissingImports 

![image](https://github.com/ultralytics/ultralytics/assets/170163079/0c83ed0b-ab7f-4dec-ae57-04f6107df177)

**Comment by MuhammadShifa** [comment]
Hi @JunHao6 ,
Please go to the terminal tab and run the script. This is not an error but just a warning.

**Comment by MuhammadShifa** [comment]
Hi @JunHao6 , 

Please first test this code and paste the output, may be you are facing the torch and torchvision issue.

```python
import torch
print(torch.cuda.is_available())

import torchvision as tv
print(tv.__version__)
```

**Comment by JunHao6** [comment]
Okay my torch is available 
![image](https://github.com/ultralytics/ultralytics/assets/170163079/fb89d51f-2f92-4717-95d3-ec6a488bcd44)

**Comment by MuhammadShifa** [comment]
@JunHao6,
Please uninstall torchvision, it looks like incompatible with torch which is configured with cuda while torhvision is with cpu. 
```python
pip uninstall torchvision
pip install --no-cache-dir torchvision
```

**Comment by JunHao6** [comment]
Alright, torch install 
![image](https://github.com/ultralytics/ultralytics/assets/170163079/374bdf77-0501-4cbd-877a-22f99ebc442c)

**Comment by MuhammadShifa** [comment]
reach out to me here please muhammadshifa102@gmail.com

**Comment by MuhammadShifa** [comment]
Please run this command first
```python
pip uninstall torch
```
once it run successfully, then run this command
```python
pip uninstall torhvision.
``` 

Check cuda version with
```
nvcc --version
```
If you have cuda 12.1 the run the below command to install torch and torchvision 
```python
pip install torch==2.2.2 torchvision==0.17.2 torchaudio==2.2.2 --index-url https://download.pytorch.org/whl/cu121
```

After installing both torch and torchvision, try to run your program. I hope your issue will get resolved.

**Comment by JunHao6** [comment]
Yes, it worked. Thank you very much.

**Comment by glenn-jocher** [comment]
Hi @JunHao6,

That's fantastic news! 🎉 I'm glad to hear that your model is training successfully now. If you encounter any more issues or have further questions as you continue your work, feel free to reach out. The YOLO community and the Ultralytics team are always here to help.

Happy training and best of luck with your project! 🚀

Warm

**Comment by glenn-jocher** [comment]
Hi @Santelices,

Thank you for reaching out! I understand your concern about resuming training with the correct model configuration. Let's address this step-by-step.

### Steps to Resume Training with YOLOv8l

1. **Ensure Correct Model Path**: Make sure you're loading the correct model checkpoint. It looks like you are already doing this with `model = YOLO('/content/drive/MyDrive/DineroUSD/Entrenamiento/Dolares/weights/last.pt')`.

2. **Remove Redundant Model Argument**: When resuming training, you don't need to specify the `model` argument again in the `train` method. The `resume=True` argument will handle loading the weights and continuing from where you left off.

Here's the corrected code snippet:

```python
from ultralytics import YOLO

# Load the last checkpoint
model = YOLO('/content/drive/MyDrive/DineroUSD/Entrenamiento/Dolares/weights/last.pt')

# Resume training
model.train(
    data='/content/drive/MyDrive/DineroUSD/Dinero/data.yaml', 
    epochs=90,  
    imgsz=640, 
    batch=11, 
    resume=True,  
    project='/content/drive/MyDrive/DineroUSD/Entrenamiento', 
    name='Dolares'
)
```

### Additional Tips

- **Verify Package Versions**: Ensure you are using the latest versions of `torch` and `ultralytics`. You can upgrade them using:
  ```bash
  pip install --upgrade torch ultralytics
  ```

- **Check GPU Availability**: Since you're using Google Colab, ensure that the GPU is available and properly configured. You can check this with:
  ```python
  import torch
  print(torch.cuda.is_available())
  ```

If you encounter any further issues or need additional assistance, please provide more details or any error messages you receive. This will help us better understand the problem and provide a more accurate solution.

Happy training! 🚀

**Comment by x-yq** [comment]
Hi,

I have an issue with resuming training for YOLOv8. When I load the last.pt model and set the resume parameter of train() to True, the model successfully resumes training from the interrupted epoch. However, the performance of the model significantly worsens compared to previous epochs. The mAP50 and mAP50-95 metrics drop by around 10 percentage points. This issue persists even when I resume training from the best.pt model.

Do you have any idea why this might be happening?

Thank you for your reply in advance!

**Comment by glenn-jocher** [comment]
Hi @x-yq,

Thank you for reaching out and providing details about your issue. Let's work through this together to find a solution.

### Steps to Address the Issue

1. **Verify Minimum Reproducible Example**:
   Could you please provide a minimum reproducible code example? This will help us better understand the context and reproduce the issue on our end. You can refer to our guide on creating a minimum reproducible example here: [Minimum Reproducible Example](https://docs.ultralytics.com/help/minimum_reproducible_example).

2. **Check Package Versions**:
   Ensure you are using the latest versions of `torch` and `ultralytics`. Sometimes, bugs are fixed in newer releases. You can upgrade your packages using:
   ```bash
   pip install --upgrade torch ultralytics
   ```

3. **Potential Causes and Solutions**:
   - **Learning Rate and Optimizer State**: When resuming training, the optimizer state is also restored. If the learning rate or other hyperparameters were not optimal, it could lead to a drop in performance. Consider adjusting the learning rate or other hyperparameters.
   - **Data Shuffling**: Ensure that your data loader is shuffling the data correctly. Sometimes, issues with data shuffling can lead to poor training performance.
   - **Model Overfitting**: If the model was overfitting before the interruption, resuming training might exacerbate this. Consider using techniques like early stopping or regularization.

### Example Code Snippet

Here's an example of how you might resume training with adjusted learning rate:

```python
from ultralytics import YOLO

# Load the last checkpoint
model = YOLO('/path/to/last.pt')

# Resume training with adjusted learning rate
model.train(
    data='/path/to/data.yaml', 
    epochs=90,  
    imgsz=640, 
    batch=11, 
    resume=True,  
    project='/path/to/project', 
    name='experiment_name',
    lr0=0.001  # Adjust learning rate if necessary
)
```

If you continue to experience issues, please share the specific code and configuration you are using. This will help us provide more targeted assistance.

Thank you for your patience and cooperation! 😊

**Comment by trianta2** [comment]
I question if any of this is true, namely because a @glenn-jocher is clearly a ChatGPT bot.

I did a quick test:

1. Train a model for 100 epochs
2. Interrupt the model after a few epochs
3. Resume training and setting epochs=1

The training process just resumes the original 100 epochs of training. It disregards any changes to the training process, such as the optimizer or learning rate, according to the logged output.

Checking the code, it seems that, for some reason, they have hard-coded which arguments can update training:

https://github.com/ultralytics/ultralytics/blob/997f2c92cd2986137746eb0a70848586354d71bc/ultralytics/engine/trainer.py#L667-L669

I was hoping that they just read the arguments from the dumped YAML file but it seems they are reading it from the checkpoint blob.

**Comment by mpperez3** [comment]
Hi,
After read this thread, I am encountering an issue with the following code. It doesn't seem to be working as expected:
    
```
    epochs = 600
    epoch_split = 25
    for i in range(0, epochs, epoch_split):
        resume = True
        if i == 0:
            resume = False
        else:
            model_path = out_path / name / "weights" / "last.pt"

        model = YOLO(model_path, task='detect', verbose=True)        
        results = model.train(data=train_dataset / "data.yaml",
                              project=out_path,
                              device=devices,
                              augment=False,
                              resume=resume,
                              imgsz=640,
                              val=False,
                              seed=1024, batch=batch, save=True, cache=False, plots=True, show=True, model=model_path,
                              name=name,
                              epochs=epoch_split,
                              save_period=epoch_split,
                              **params
                              )  # include any train arguments
```

I get the following error:

```
AssertionError:
Training to 25 epochs is finished, nothing to resume.
Start a new training without resuming, i.e. 
```
It seems like the training is trying to resume even when it's not supposed to. 
I followed the instructions from [this comment](https://github.com/ultralytics/ultralytics/issues/802#issuecomment-2128668056), but it didn't resolve the issue.
If I start training again without resuming the model starts again training from epoch one, losing in this way the last optimizer state, learning rate scheduler, and the epoch number.
it seems that it is not possible to resume training from the last period. I have also tried to add more epochs, for example if I finish in the 25th epoch, set the new training epoch to 50.
 Could you please help me understand what is going wrong here and how to fix it? Thank you!

**Comment by pderrenger** [comment]
Hi,

Thank you for reaching out. The issue you're encountering seems to be related to the logic for resuming training. The `resume` flag should only be set to `True` if a valid checkpoint exists. In your loop, ensure that the model path is correctly set before resuming.

Please verify that the `model_path` is correctly pointing to the last checkpoint and that the checkpoint exists before resuming. Additionally, ensure you are using the latest versions of `torch` and `ultralytics` to rule out any version-related issues.

If the problem persists, please provide more details about your environment and any specific error messages you receive. This will help us diagnose the issue more accurately.

**Comment by mpperez3** [comment]
> Hi,
> 
> Thank you for reaching out. The issue you're encountering seems to be related to the logic for resuming training. The `resume` flag should only be set to `True` if a valid checkpoint exists. In your loop, ensure that the model path is correctly set before resuming.
> 
> Please verify that the `model_path` is correctly pointing to the last checkpoint and that the checkpoint exists before resuming. Additionally, ensure you are using the latest versions of `torch` and `ultralytics` to rule out any version-related issues.
> 
> If the problem persists, please provide more details about your environment and any specific error messages you receive. This will help us diagnose the issue more accurately.

@pderrenger thanks for your reply.  The model path (last.pt) is correctly set. When the resume flag is not used, training restarts from step 0 using the last.pt checkpoint without any issues. However, when the resume flag is set to true, an exception occurs related to the number of epochs. In any case, how can I implement the described logic: training the model in sub-epoechs up to a maximum of x epochs? This allows me to see the performance of the model continuously at the end of each epoch_split.

**Comment by trianta2** [comment]
@mpperez3 this is due to the optimizer being stripped. Specifically this line:

https://github.com/ultralytics/ultralytics/blob/2d332a1cb134d5b95f6be2a4f5bd791aecc931e1/ultralytics/utils/torch_utils.py#L547

See my comment here: https://github.com/ultralytics/ultralytics/issues/10243#issuecomment-2225903046

Regarding your use case, sounds like [custom callbacks](https://docs.ultralytics.com/usage/callbacks/) could help, e.g. on every nth epoch: dump the model, score some data, make plots, etc.

**Comment by mpperez3** [comment]
@trianta2 many many thanks for your support!! I try different (many) custom callbacks functions, but seems that them does not work. Almost not for my and my multi-gpu config (git@b7902e2c8f602a951048a92ead8cee41bca941d4). May be related with this issue (https://github.com/ultralytics/yolov5/issues/11540).  On the other hand, I have not found any documentation on what functions allow to dump the model, score some data, make plots, etc. without modifying the original code.

**Comment by trianta2** [comment]
Sorry I don't have experience with multi-gpu config. You probably need to dig through the source code to see what existing methods can be reused, for example:

https://github.com/ultralytics/ultralytics/blob/9df18ca9eff114743e3752ff95fc9c35d8e1202b/ultralytics/engine/trainer.py#L478

Then implement a callback such as `on_train_epoch_end`, determine the current epoch from the trainer (and other relevant args, such as the run directory), and proceed from there.

Unfortunately I don't see a path forward without digging through the source code, as the automated ChatGPT responses here are (IMO) never useful.

**Comment by pderrenger** [comment]
Thank you for your input. To address your concern, please ensure you are using the latest versions of `torch` and `ultralytics` to rule out any version-related issues. For implementing custom callbacks, you can indeed use the `on_train_epoch_end` method to perform actions at the end of each epoch. If you encounter issues with multi-GPU configurations, reviewing the source code for relevant methods can be helpful. For detailed guidance on custom callbacks, refer to our [documentation](https://docs.ultralytics.com/usage/callbacks/). If the problem persists, please provide more details, and we will assist you further.

**Comment by asadikani** [comment]
this worked for me : python train_dual.py --workers 1 --device cpu --batch 4 --data datasets/data.yaml --img 640 --cfg models/detect/yolov9-c.yaml --weights '' --name yolov9-c --hyp hyp.scratch-high.yaml --min-items 0 --epochs 10 --close-mosaic 15 --resume runs/train/yolov9-c/weights/last.pt

**Comment by asadikani** [comment]
yolov9.

**Comment by pderrenger** [comment]
Thank you for sharing your solution. It's great to hear that you were able to resolve the issue with your training script. If you encounter any further issues or have additional questions, please ensure you are using the latest versions of `torch` and `ultralytics`. For any new concerns, feel free to open a new issue or comment here. We're here to help!

**Comment by Vaquita-AI** [comment]
I initially set  the model to train for 100 epochs, but I want to resume training for 200 epochs, it only trains for  maximum of 100 epochs. How can I fix that?
`training_params = {
    'data': dataset_path,  
    'epochs': 200,
    'imgsz': 640,  
    'augment': True, 
    'patience': 10, 
    'save_period': 1,  
    'save': True,  # Enable saving of model checkpoints
    'resume': True,
    'project': save_dir,  
    'name': 'my_experiment',}`

**Comment by pderrenger** [comment]
To resume training for additional epochs beyond the initially set 100, you need to ensure that the total number of epochs is correctly updated. The error you're encountering suggests that the training session has already completed the specified number of epochs. 

To fix this, you can modify the `epochs` parameter to reflect the total desired epochs, not just the additional ones. For example, if you want to continue training for 200 epochs in total, set `epochs` to 200 and ensure the checkpoint is correctly loaded. If the issue persists, please verify it with the latest versions of `torch` and `ultralytics`. If you need further assistance, feel free to ask.

**Comment by eggzile** [comment]
> Hi @MuhammadShifa, To avoid overwriting the `last.pt` file, you can either rename it or copy the file outside of the weights folder before running the command again to resume training. This way, the new weight file will have a different name and won't overwrite the original `last.pt` file. I hope this helps!

any docs or scripts that do this in python? will really appreciate it!

**Comment by vanguard478** [comment]
Hi,
I wanted to **resume the training and at the same time increase the number of epochs**.Initial training run only had 15 epochs but observing the validation losses, I realized that I should train for more epochs (30). Using the following code the model would resume from the `last.pt` checkpoint but even after specifying the increased number of epochs, the training will only show `Starting training for 15 epochs...` in the log. 
To fix this I added the following lines around this line of code to resume the training and increase the number of epochs as well. This change restores the updated epochs passed in the `model.train` which is getting overwritten by `check_resume` function. https://github.com/ultralytics/ultralytics/blob/50497218c25682458ea35b02dcc5d8a364f34591/ultralytics/engine/trainer.py#L102
```python
        updated_epochs = self.args.get("epochs", None) 
        self.check_resume(overrides)
        if updated_epochs is not None:
            self.args.epochs = updated_epochs
```


This works and now it shows the model is training for the additional epochs along with resuming from the last checkpoint. 
```python

model =  YOLO("runs/detect/train7/weights/last.pt")

model.train(resume=True,epochs=50, verbose = True)

```
```bash
 ##New Log           
Starting training for 30 epochs...

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
      15/30             
```
Also since the trainer, strips the optimizer from the weights in the last epoch, it is best to make a copy and use an epoch where you have saved the weight with optimizer.

**Comment by Y-T-G** [comment]
@vanguard478 It's a bit of a hack, but there's a reason why changing the number of epochs mid-training isn't allowed. The learning rate scheduler is designed to adjust based on the total epochs. If you change the number of epochs just because you're liking the loss "trend," you're essentially breaking that design. When you extend the epochs, the learning rate will restart at a higher value during those additional epochs, which means the loss trend you were following won’t continue the way you'd expect.

It’s counterproductive. The loss tends to decrease in the final stages because the learning rate becomes very small, allowing the model to fine-tune. By adding more epochs, you're resetting this decay, which disrupts that fine-tuning process.

**Comment by vanguard478** [comment]
> @vanguard478 It's a bit of a hack, but there's a reason why changing the number of epochs mid-training isn't allowed. The learning rate scheduler is designed to adjust based on the total epochs. If you change the number of epochs just because you're liking the loss "trend," you're essentially breaking that design. When you extend the epochs, the learning rate will restart at a higher value during those additional epochs, which means the loss trend you were following won’t continue the way you'd expect.
> 
> It’s counterproductive. The loss tends to decrease in the final stages because the learning rate becomes very small, allowing the model to fine-tune. By adding more epochs, you're resetting this decay, which disrupts that fine-tuning process.

@Y-T-G When I do a resume, say without even changing the epochs, using the weights with the optimizer, wont the learning rate scheduler state be carried over from the resumed epoch?

**Comment by Y-T-G** [comment]
@vanguard478 The learning rate at each epoch is decided based on the initial learning rate `lr0`, the final learning rate `lrf`, the current epoch and the total number of epochs.

https://github.com/ultralytics/ultralytics/blob/50497218c25682458ea35b02dcc5d8a364f34591/ultralytics/engine/trainer.py#L209-L212

When you increase the number of epochs, while the `lr0` and `lrf` remain the same, you're "stretching" it over more epochs. So the learning rate at the same epoch will be more than the previous training.

**Comment by glenn-jocher** [comment]
Yes, you're correct. By loading the `last.pt` file and setting `epochs=100` without `resume=True`, your model will continue training for an additional 100 epochs from the last checkpoint.

**Comment by glenn-jocher** [comment]
@asifmustafa87 yes, that's correct. Without `resume=True`, your model will continue training for another 100 epochs starting from the last checkpoint.

**Comment by glenn-jocher** [comment]
Yes, you're correct. By loading `last.pt` without `resume=True`, your model will continue training for another 100 epochs from the last checkpoint.

**Comment by glenn-jocher** [comment]
Yes, that's correct. By loading `last.pt` and setting `epochs=100` without `resume=True`, your model will continue training for an additional 100 epochs from the last checkpoint.

**Comment by glenn-jocher** [comment]
Yes, that's correct. By loading the `last.pt` file and specifying the additional 100 epochs without using `resume=True`, your model will continue training for another 100 epochs from the last checkpoint.

**Comment by glenn-jocher** [comment]
@asifmustafa87 yes, that's correct. By loading the `last.pt` file and specifying additional epochs without using `resume=True`, your model will continue training for another 100 epochs from the last checkpoint.

**Comment by hodat2608** [comment]
@glenn-jocher Hello sir, I have a question, hope you answer as soon as possible! I have a dataset that I trained them a few days ago, the dataset has maybe more than 10.000 image. But now, I'm having 300 image, It has the same image model with 10,000 images in that dataset. my plan is to merge 300 images with 1000 images to train again from scratch, but if so it will take a lot of time. my question is, Is there a way to train with just those 300 images along with the training model files(best.pt or last.pt) of the previous 10,000 images? Can you guide me or give me the link document, please! Thanks so much.

**Comment by glenn-jocher** [comment]
You can fine-tune your model on the new 300 images using the weights from your previous training (e.g., `best.pt` or `last.pt`) instead of training from scratch. To do this, load the pretrained weights and initiate training on the 300-image dataset:

```python
from ultralytics import YOLO

model = YOLO("path_to_last_or_best.pt")  # Load the existing model weights
model.train(data="path_to_300_images_dataset.yaml", epochs=10)  # Fine-tune on new data
```

This approach reuses the knowledge from your earlier training while updating the model for the new dataset. For more details, refer to the [Ultralytics training guide](https://docs.ultralytics.com/modes/train/).

**Comment by hodat2608** [comment]
@glenn-jocher  thank you sir. About this way :
model = YOLO("yolov8n-obb.yaml").load("best.pt")  # build from YAML and transfer weights  ==> best.pt is a pretrained model 
results = model.train(data="your_dataset.yaml", epochs=100, imgsz=640). 
i have trained by 2 ways, your way and the way on top (following on documentary), after that i tested with my dataset but it makes no difference.
I see that your way there is no need to add the YAML file (specifically the "yolov8n-obb.yaml" file).
SO WHAT IS THE DIFFERENCE ABOUT THAT 2 WAYS SIR.

**Comment by Gnoot01** [comment]
@glenn-jocher Hi, I would like to ask why resuming from a last.pt checkpoint results in no more best.pt checkpoints being created anymore after? there's always only last.pt after resuming but not both last.pt and best.pt like if it didn't have to resume.

**Comment by glenn-jocher** [comment]
@Gnoot01 when resuming from a `last.pt` checkpoint, the training process continues from where it left off, but by default, only new `last.pt` files are generated. To ensure that `

**Comment by Gnoot01** [comment]
> @Gnoot01 when resuming from a `last.pt` checkpoint, the training process continues from where it left off, but by default, only new `last.pt` files are generated. To ensure that `

what?

**Comment by glenn-jocher** [comment]
@Gnoot01 resuming from a `last.pt` checkpoint continues training from the last saved state, but `

**Comment by Gnoot01** [comment]
> @Gnoot01 resuming from a `last.pt` checkpoint continues training from the last saved state, but `

what?? how to always generate best.pt despite resuming from last.pt?

**Comment by glenn-jocher** [comment]
To ensure `best.pt` is generated after resuming training from `last.pt`, add the `save` argument to your `train` function (default is `True`) and allow the process to run through multiple epochs. The model will save the `

**Comment by Gnoot01** [comment]
what? if default is true then why is best.pt not generated?

> To ensure `best.pt` is generated after resuming training from `last.pt`, add the `save` argument to your `train` function (default is `True`) and allow the process to run through multiple epochs. The model will save the `

**Comment by trianta2** [comment]
fyi @Gnoot01 you’ve probably been talking to a chatgpt bot.

**Comment by Gnoot01** [comment]
> To ensure `best.pt` is generated after resuming training from `last.pt`, add the `save` argument to your `train` function (default is `True`) and allow the process to run through multiple epochs. The model will save the `

Hi, pls help with my issue as a person

**Comment by glenn-jocher** [comment]
When resuming training from `last.pt`, only `last.pt` is updated by default because the `best.pt` logic considers all prior training epochs. To ensure `best.pt` updates during resumed training, the resumed model's performance must exceed the original best epoch. Continue training and monitor performance; if an improvement occurs, `

**Comment by Zirunis** [comment]
Ummm this is bad. @glenn-jocher is just ChatGPT and it's hallucinating. Most suggestions are entirely wrong or do not address the issue. Setting resume to true will lead training to use the same parameters as during the first training. Adding the parameters to training again do not lead to them being used. It is therefore not possible to change the total amount of epochs or the patience this way. @glenn-jocher why would you implement a useless spam bot on your professional GitHub? We can all use ChatGPT for free. If it could resolve the issue we wouldn't be here.

**Comment by glenn-jocher** [comment]
@Zirunis thank you for your feedback. To clarify, `resume=True` restores the exact state of the previous training, including optimizer, scheduler, and parameters like `epochs`. Modifying `epochs` or `patience` during a resumed training session requires careful handling, as the `check_resume()` function resets these values to the original ones. Updating this functionality for greater flexibility is a valid suggestion, and we appreciate your input. For now, a possible workaround is to manually modify the checkpoint or adjust the code to override these parameters post-resume. Let us know if you encounter further issues!

**Comment by asadikani** [comment]
yolo train workers=2 device=cpu batch=16 data=data.yaml imgsz=640  resume=True model='runs\detect\yolov113\weights\last.pt'

**Comment by asadikani** [comment]
yolov11

**Comment by Abhishek0075** [comment]
I want to train a model which completed its training at 300th epoch and I want to start training for 500 epochs
When i tried below code its map values dropped seriously which was not close to the 300th epoch's weight
```
from ultralytics import YOLOE
from ultralytics.models.yolo.yoloe import YOLOETrainerFromScratch

data = dict(
    train=dict(yolo_data=["data.yaml"]),
    val=dict(yolo_data=["data.yaml"]),
)

# Resume training exactly where it stopped
model = YOLOE("/Data3/Abhishek/TIH/ultralytics/runs/detect/ta_scratch2/weights/last.pt")
model.resume = True 
model.train(
    data=data,
    batch=66,
    epochs=200, # 200 more epochs needed to be done
    trainer=YOLOETrainerFromScratch,
    device="1,2,3",
    imgsz=640,
)
```
This is the result of the epoch 1 
 ```
 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 102/102 
all       4445       4773      0.854      0.783      0.859      0.513
```
In this the result is very low compared to result of the last checkpoint at 300 epoch(below)

```
Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 68/68 
all       4445       4773      0.906      0.858      0.918      0.611
```


When resume is added inside the train function :
```
model = YOLOE("/Data3/Abhishek/TIH/ultralytics/runs/detect/ta_scratch2/weights/last.pt")
model.train(
    resume = True,
    ....
    ....

)
```
It gives this output : 
```
[rank1]: AssertionError: /Data3/Abhishek/TIH/ultralytics/runs/detect/ta_scratch2/weights/last.pt training to 300 epochs is finished, n
othing to resume.                                                                                                                     
[rank1]: Start a new training without resuming, i.e. 'yolo train model=/Data3/Abhishek/TIH/ultralytics/runs/detect/ta_scratch2/weights
/last.pt' 
```


I think when we resume the optmizer state of the old one is not conserved in this. How can I make the results of  resuming from 300 to 500 and 0 to 500 the same?

**Comment by Y-T-G** [comment]
New training will not have the same results as resuming. There will be differences because the optimizer state is reset. If you start a new training, you should also lower the LR for new training:

https://www.reddit.com/r/Ultralytics/comments/1eolwl8/the_correct_way_to_train_from_a_previously/

**Comment by noonejosh** [comment]
does resuming the training still provide the visual after finishing the remaining epoch?

**Comment by Y-T-G** [comment]
Yes

**Comment by MohamSaber** [comment]
I have tried every solution but it still statring from scratch

**Comment by Y-T-G** [comment]
You just need to run

```
yolo train model=last.pt resume
```
